File: p4.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <ctype.h>

int precedence(char op) {
    if (op == '^') return 3;
    if (op == '*' || op == '/') return 2;
    if (op == '+' || op == '-') return 1;
    return 0;
}

int isRightAssociative(char op) {
    return op == '^';
}

char* infixToPostfix(char* s) {
    int len = strlen(s);
    char* result = (char*)malloc((len + 1) * sizeof(char));
    char* stack = (char*)malloc(len * sizeof(char));
    int top = -1, k = 0;

    for (int i = 0; i < len; i++) {
        char ch = s[i];
        if (isalnum(ch)) {
            result[k++] = ch;
        } else if (ch == '(') {
            stack[++top] = ch;
        } else if (ch == ')') {
            while (top != -1 && stack[top] != '(') {
                result[k++] = stack[top--];
            }
            top--;
        } else {
            while (top != -1 && ((precedence(stack[top]) > precedence(ch)) ||
                   (precedence(stack[top]) == precedence(ch) && !isRightAssociative(ch)))) {
                if (stack[top] == '(') break;
                result[k++] = stack[top--];
            }
            stack[++top] = ch;
        }
    }

    while (top != -1) {
        result[k++] = stack[top--];
    }

    result[k] = '\0';
    free(stack);
    return result;
}

File: p4.py
class Solution:
    def infixtoPostfix(self, s):
        def precedence(op):
            if op == '^':
                return 3
            if op in ('*', '/'):
                return 2
            if op in ('+', '-'):
                return 1
            return 0

        def isRightAssociative(op):
            return op == '^'

        stack = []
        result = []

        for ch in s:
            if ch.isalnum():
                result.append(ch)
            elif ch == '(':
                stack.append(ch)
            elif ch == ')':
                while stack and stack[-1] != '(':
                    result.append(stack.pop())
                stack.pop()
            else:
                while (stack and precedence(stack[-1]) > precedence(ch)) or \
                      (stack and precedence(stack[-1]) == precedence(ch) and not isRightAssociative(ch)):
                    if stack[-1] == '(':
                        break
                    result.append(stack.pop())
                stack.append(ch)

        while stack:
            result.append(stack.pop())

        return "".join(result)
